codejar.ts

```diff
@@ -121,23 +126,38 @@ export function CodeJar(editor: HTMLElement, highlight: (e: HTMLElement, pos?: P
     const s = window.getSelection()!
     const pos: Position = {start: 0, end: 0, dir: undefined}
 
+    let {anchorNode, anchorOffset, focusNode, focusOffset} = s
+    // selection anchor and focus are expected to be text nodes, so normalize them
+    if (anchorNode.nodeType === Node.ELEMENT_NODE) {
+      const node = document.createTextNode('')
+      anchorNode.insertBefore(node, anchorNode.childNodes[anchorOffset])
```

Comment:
```
anchorNode.childNodes[anchorOffset]
```
This code is strange, childNodes is a list of elements and anchorOffset is a character offset. From my understanding it's not possible to use it this way.

Reply:
The meaning of `anchorOffset` is overloaded depending on the `nodeType` of the `anchorNode`.

[Per MDN](https://developer.mozilla.org/en-US/docs/Web/API/Selection#properties) on `anchorOffset`, emphasis mine:

> Returns a number representing the offset of the selection's anchor within the `anchorNode`. If `anchorNode` is a text node, this is the number of characters within `anchorNode` preceding the anchor. **If `anchorNode` is an element, this is the number of child nodes of the `anchorNode` preceding the anchor.**

On Chrome, this appears to only have an impact when modifying those properties. (`anchorNode` is only ever set to a text node by the browser), but Firefox makes use of this overloaded definition, and sometimes sets the `anchorNode` to an element, setting the `anchorOffset` appropriately.

(Analogously for `focusNode` and `focusOffset`.)

This is per my third bullet item:

> Normalize the selection before using it. **Firefox will occasionally report elements (rather than text nodes) as selection endpoints** […]

I suppose the comment I wrote wasn’t clear enough, though. What I meant was that *the following code* expects for the anchor and focus to be text nodes. (Which is not necessarily the case in Firefox.) I can rewrite the comment if you think it’s important enough.

Reply:
Got it. Yes, please, update the comment.

Reply:
Also is there a way to test this issue? Is there some reliable way in firefox to force the selection to be on an element node?